packages/ketcher-react/src/script/editor/utils/canvasExtension.ts

```diff
@@ -0,0 +1,209 @@
+    horizontalExtension = Math.abs(newHorizontalScrollPosition);
+  }
+  if (newVerticalScrollPosition > currentCanvasSize.y) {
+    verticalExtension = newVerticalScrollPosition - currentCanvasSize.y;
+  }
+  if (newVerticalScrollPosition < 0) {
+    verticalExtension = Math.abs(newVerticalScrollPosition);
+  }
+  return new Vec2(horizontalExtension, verticalExtension, 0);
+}
+
+export function shiftByVector(vector: Vec2, editor: Editor) {
+  const render = editor.render;
+  const clientArea = render.clientArea;
+  const extensionVector = calculateCanvasExtension(
+    clientArea,
+    render.sz.scaled(render.options.zoom),
+    vector,
+  ).scaled(1 / render.options.zoom);
+
+  if (extensionVector.x > 0 || extensionVector.y > 0) {
+    // When canvas extends previous (0, 0) coordinates may become (100, 100)
+    lastX += extensionVector.x;
+    lastY += extensionVector.y;
+  }
+  render.update(false);
+  scrollToEdgeOfScreen(vector, render);
+}
+
+export function scrollByVector(vector: Vec2, render) {
+  requestAnimationFrame(() => {
+    const clientArea = render.clientArea;
+    clientArea.scrollLeft += vector.x * scrollMultiplier;
+    clientArea.scrollTop += vector.y * scrollMultiplier;
+  });
+}
+
+export function scrollToEdgeOfScreen(vector: Vec2, render) {
+  requestAnimationFrame(() => {
+    const clientArea = render.clientArea;
+    if (vector.x < 0) {
+      clientArea.scrollLeft = 0;
+    }
+    if (vector.x > 0) {
+      clientArea.scrollLeft = clientArea.scrollWidth;
+    }
+    if (vector.y < 0) {
+      clientArea.scrollTop = 0;
+    }
+    if (vector.y > 0) {
+      clientArea.scrollTop = clientArea.scrollHeight;
+    }
+  });
+}
+
+export function scrollByDirection(editor: Editor, direction) {
+  const clientArea = editor.render.clientArea;
+  const selectedAreaCoordinates = getSelectedAreaCoordinates(editor);
+  if (!selectedAreaCoordinates) {
+    return;
+  }
+  const { leftX, topY } = selectedAreaCoordinates;
+  if (direction === 'MoveUp') {
+    clientArea.scrollTop = topY - 10;
+  } else if (direction === 'MoveLeft') {
+    clientArea.scrollLeft = leftX - 10;
+  } else if (direction === 'MoveRight') {
+    clientArea.scrollLeft += scrollMultiplier;
+  } else if (direction === 'MoveDown') {
+    clientArea.scrollTop += scrollMultiplier;
+  }
+}
+
+export function setScrollOnSelection(editor: Editor, direction) {
+  const clientArea = editor.render.clientArea;
+  const selectedAreaCoordinates = getSelectedAreaCoordinates(editor);
+  if (!selectedAreaCoordinates) {
+    return;
+  }
+  const { leftX, topY } = selectedAreaCoordinates;
+  if (direction === 'MoveUp') {
+    clientArea.scrollTop = topY - 10;
+  } else if (direction === 'MoveLeft') {
+    clientArea.scrollLeft = leftX - 10;
+  } else if (direction === 'MoveRight') {
+    clientArea.scrollLeft = leftX - clientArea.clientWidth / 2;
+  } else if (direction === 'MoveDown') {
+    clientArea.scrollTop = topY - clientArea.clientHeight / 2;
+  }
```

Comment:
When structure is near the edge of screen and is moved out of visible area,  clientArea scroll may not keep up with the movement of the structure, so when mouse is released, we set scroll to make structure visible.  ToDo: Use clientArea.scrollTo with behavior: 'smooth' here.